Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,242 @@
+       * otherwise, it's set to false.
+       * @param model The model to associate.
+       * @param value The associated value.
+       */
+      void associate(const std::shared_ptr<BooleanModel>& model,
+        const Type& value);
+
+      /**
+       * Constructs a LocalBooleanModel and associates it with the given
+       * value.
+       * @param value The associated value.
+       * @return The associated model.
+       */
+      std::shared_ptr<BooleanModel> make_association(const Type& value);
+
+      /**
+       * Finds the model associated with a value.
+       * @param value The model's associated value.
+       * @return The associated model, or nullptr if no associated model was
+       *         found.
+       */
+      std::shared_ptr<BooleanModel> find(const Type& value) const;
+
+      /**
+      * Sets which value should become current if all associated models have
+      * a value of false.
+      * @param value The default value.
+      * @return Acceptable if default value was set successfully, Invalid
+      *         otherwise.
+      */
+      QValidator::State set_default_value(const Type& value);
+
+      /**
+       * Returns Acceptable if the model's current value is valid, Invalid
+       * otherwise.
+       * @return The state of the AssociativeValueModel.
+       */
+      QValidator::State get_state() const override;
+
+      /**
+       * Returns the AssociativeValueModel's current value.
+       * @return The current value.
+       */
+      const Type& get_current() const override;
+
+      /**
+       * Sets the current value, iff a model is associated with the value.
+       * @param value The current value.
+       * @return Acceptable if the value was set successfully, Invalid
+       *         otherwise.
+       */
+      QValidator::State set_current(const Type& value) override;
+
+      /** Connects a slot to the current signal. */
+      boost::signals2::connection connect_current_signal(
+        const typename CurrentSignal::slot_type& slot) const override;
+
+    private:
+      mutable CurrentSignal m_current_signal;
+      Type m_current;
+      std::unique_ptr<Type> m_default_value;
+      std::unordered_map<Type, Details::AssociatedBooleanModel, Hash> m_models;
+      bool m_is_blocked;
+
+      void set_associated_model_value(const Type& value, bool model_value);
+      void on_current(const Type& value, bool is_selected);
+  };
+
+  template<typename T, typename U>
+  AssociativeValueModel<T, U>::AssociativeValueModel()
+    : m_is_blocked(false) {}
+
+  template<typename T, typename U>
+  void AssociativeValueModel<T, U>::associate(
+      const std::shared_ptr<BooleanModel>& model, const T& value) {
+    if(m_models.find(value) != m_models.end()) {
+      return;
+    }
+    if(model->get_current()) {
+      model->set_current(false);
+    }
+    m_models.insert_or_assign(value, Details::AssociatedBooleanModel{model,
```

Comment:
Reuse the iterator from line 114.

Reply:
The iterator on line 114 would equal m_models.end() if it was used. I can use the plain insert method: m_models.insert({value, model}), but I don't see the advantage of using the iterator.

Reply:
Yes good point. What you want to do is along the lines of:

    auto i = m_models.insert({value, model});
    if(i.second) {
      return;
    }
    ...